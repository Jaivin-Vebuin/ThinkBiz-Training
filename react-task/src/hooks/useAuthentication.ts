import { useDispatch } from "react-redux";
import { loginAPI, registerAPI } from "../services/userAuthServices";
import { setAuthToken } from "../redux/features/slices/auth/authSlice";
import {
  loginDataType,
  registerDataType,
} from "../data/model/types/hooksTypes/useAuthentication";
import { toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";
import { useNavigate } from "react-router";
import { useTranslation } from "react-i18next";

function useAuthentication() {
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const {t} = useTranslation();

  const registerUser = async (data: registerDataType) => {
    try {
      
      await registerAPI(data);
      navigate('/login')
      toast.success(t("registrationToastSuccess"), {
        position: "top-center",
      });
    } catch (error) {
      if (error instanceof Error) {
        toast.error(`${t("registrationToastError")}: ${error.message}`, {
          position: "top-center",
        });
      }
    }
  };

  const loginUser = async (data: loginDataType) => {
    try {
   
      const loginResponse = await loginAPI(data);
      const token = loginResponse.data;
      dispatch(setAuthToken(token));
      if (loginResponse.status === 200) {
       
        toast.success(t("loginToastSuccess"), { position: "top-center" });
      } else {
        toast.error(`${t("loginToastError")}: ${loginResponse.data.message}`, {
          position: "top-center",
        });
      }
    } catch (error) {
      if (error instanceof Error) {
        toast.error(`${t("loginToastError")}: ${error.message}`, {
          position: "top-center",
        });
      }
    }
  };

  return { registerUser, loginUser };
}

export default useAuthentication;
