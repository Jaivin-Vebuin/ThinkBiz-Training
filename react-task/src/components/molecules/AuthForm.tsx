import React from "react";
import Input from "../atoms/Input";
import Select from "../atoms/Select";
import { roles } from "../../utils/constants/app.constants";
import {
  FormStyle,
  TitleStyle,
  InputFieldStyle,
  ButtonStyle,
  FormRedirectionButton,
} from "../styles/AuthFormStyles";
import { useForm, SubmitHandler } from "react-hook-form";
import { joiResolver } from "@hookform/resolvers/joi";
import {
  registerSchema,
  loginSchema,
} from "../../data/validation/formValidation";
import Button from "../atoms/Button";
import useAuthentication from "../../hooks/useAuthentication";
import { useTranslation } from "react-i18next";
import { useNavigate } from "react-router";

interface formPropsType {
  formType: string;
}

const AuthForm: React.FC<formPropsType> = ({ formType }) => {
  const navigate = useNavigate();
  const { t } = useTranslation();
  const { loginUser, registerUser } = useAuthentication();

  const schema = formType === "register" ? registerSchema : loginSchema;
  const {
    register,
    handleSubmit,
    formState: { errors, touchedFields },
  } = useForm({
    resolver: joiResolver(schema),
  });

  const onSubmit: SubmitHandler<any> = (data) => {
    if (formType === "login") {
      loginUser({ email: data.userEmail, password: data.userPassword });
    } else {
      registerUser({
        name: data.userName,
        email: data.userEmail,
        password: data.userPassword,
        role: data.userRole,
        age: parseInt(data.userAge),
      });
    }
  };

  return (
    <>
      <FormStyle onSubmit={handleSubmit(onSubmit)}>
        <TitleStyle>
          {formType === "login" ? t("formTitleLogin") : t("formTitleRegister")}
        </TitleStyle>
        <div>
          {formType === "register" && (
            <>
              <InputFieldStyle>
                <Input
                  name="userName"
                  handleOnChange={() => {}}
                  inputType="text"
                  placeholder={t("namePlaceHolder")}
                  labelText={t("Name")}
                  htmlFor="userName"
                  registerProps={register("userName")}
                  error={
                    touchedFields.userName
                      ? (errors.userName?.message as string)
                      : undefined
                  }
                />
              </InputFieldStyle>
              <InputFieldStyle>
                <Input
                  name="userEmail"
                  handleOnChange={() => {}}
                  inputType="text"
                  placeholder={t("emailPlaceHolder")}
                  labelText={t("Email")}
                  htmlFor="userEmail"
                  registerProps={register("userEmail")}
                  error={
                    touchedFields.userEmail
                      ? (errors.userEmail?.message as string)
                      : undefined
                  }
                />
              </InputFieldStyle>
              <InputFieldStyle>
                <Input
                  name="userPassword"
                  handleOnChange={() => {}}
                  inputType="password"
                  placeholder="********"
                  labelText={t("Password")}
                  htmlFor="userPassword"
                  registerProps={register("userPassword")}
                  error={
                    touchedFields.userPassword
                      ? (errors.userPassword?.message as string)
                      : undefined
                  }
                />
              </InputFieldStyle>
              <InputFieldStyle>
                <Select
                  options={roles}
                  handleOnChange={() => {}}
                  name="userRole"
                  error={
                    touchedFields.userRole
                      ? (errors.userRole?.message as string)
                      : undefined
                  }
                  labelText={t("Role")}
                  registerProps={register("userRole")}
                  htmlFor="userRole"
                />
              </InputFieldStyle>
              <InputFieldStyle>
                <Input
                  name="userAge"
                  handleOnChange={() => {}}
                  inputType="number"
                  placeholder={t("agePlaceHolder")}
                  labelText={t("Age")}
                  htmlFor="userAge"
                  registerProps={register("userAge")}
                  error={
                    touchedFields.userAge
                      ? (errors.userAge?.message as string)
                      : undefined
                  }
                />
              </InputFieldStyle>
            </>
          )}

          {formType === "login" && (
            <>
              <InputFieldStyle>
                <Input
                  name="userEmail"
                  handleOnChange={() => {}}
                  inputType="text"
                  placeholder={t("emailPlaceHolder")}
                  labelText={t("Email")}
                  htmlFor="userEmail"
                  registerProps={register("userEmail")}
                  error={
                    touchedFields.userEmail
                      ? (errors.userEmail?.message as string)
                      : undefined
                  }
                />
              </InputFieldStyle>
              <InputFieldStyle>
                <Input
                  name="userPassword"
                  handleOnChange={() => {}}
                  inputType="password"
                  placeholder="********"
                  labelText={t("Password")}
                  htmlFor="userPassword"
                  registerProps={register("userPassword")}
                  error={
                    touchedFields.userPassword
                      ? (errors.userPassword?.message as string)
                      : undefined
                  }
                />
              </InputFieldStyle>
            </>
          )}
        </div>
        <ButtonStyle>
          <Button
            text={
              formType === "login"
                ? t("loginFormButton")
                : t("registerFormButton")
            }
            handleOnClick={handleSubmit(onSubmit)}
          />
        </ButtonStyle>
      </FormStyle>
      <div
        style={{
          padding: "15px",
          borderRadius: "8px",
          textAlign: "center",
          margin: "20px",
        }}
      >
        {formType === "login" && (
          <>
            {t("forNewUser")}
            <FormRedirectionButton>
              <button
                onClick={(e) => {
                  e.preventDefault();
                  navigate("/register");
                }}
              >
                {t("registerFormButton")}
              </button>
            </FormRedirectionButton>
          </>
        )}

        {formType === "register" && (
          <>
            {t("forRegisteredUser")}
            <FormRedirectionButton>
              <button
                onClick={(event) => {
                  event.preventDefault();
                  navigate("/login");
                }}
              >
                {t("loginFormButton")}
              </button>
            </FormRedirectionButton>
          </>
        )}
      </div>
    </>
  );
};

export default AuthForm;
