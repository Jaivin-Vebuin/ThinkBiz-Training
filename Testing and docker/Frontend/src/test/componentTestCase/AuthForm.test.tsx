import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import { useNavigate } from "react-router";
import AuthForm from "../../components/molecules/AuthForm";
import useAuthentication from "../../hooks/useAuthentication";

jest.mock("react-router", () => ({
  useNavigate: jest.fn(),
}));

jest.mock("react-redux", () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(() => ({})),
}));

// jest.mock("react-i18next", () => ({
//   useTranslation: () => ({ t: (key: string) => key }),
// }));

jest.mock("../../hooks/useAuthentication", () => ({
  __esModule: true,
  default: jest.fn(() => ({
    loginUser: jest.fn(),
    registerUser: jest.fn(),
  })),
}));

let mockNavigate: jest.Mock;
describe("Auth Form", () => {
  beforeEach(() => {
    mockNavigate = jest.fn();
    (useNavigate as jest.Mock).mockReturnValue(mockNavigate);
    (useAuthentication as jest.Mock).mockReturnValue({
      loginUser: jest.fn(),
      registerUser: jest.fn(),
    });
    jest.clearAllMocks();
  });

  test("matches snapshot for register form", () => {
    const { asFragment } = render(<AuthForm formType="register" />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("matches snapshot for login form", () => {
    const { asFragment } = render(<AuthForm formType="login" />);
    expect(asFragment()).toMatchSnapshot();
  });
  test("renders login form by default and redirects to dashboard on login", async () => {
    const mockLoginUser = jest.fn();
    (useAuthentication as jest.Mock).mockReturnValue({
      loginUser: mockLoginUser,
      registerUser: jest.fn(),
    });

    render(<AuthForm formType="login" />);

    expect(screen.getByRole("heading", { name: /login/i })).toBeInTheDocument();

    const emailInput = screen.getByLabelText(/email/i);
    const passwordInput = screen.getByLabelText(/password/i);
    const loginSubmitButton = screen.getByRole("button", { name: /login/i });

    await userEvent.type(emailInput, "test@example.com");
    await userEvent.type(passwordInput, "password123");

    await userEvent.click(loginSubmitButton);

    expect(mockLoginUser).toHaveBeenCalledWith({
      email: "test@example.com",
      password: "password123",
    });
  });

  test("navigates to register form when clicking register button", async () => {
    render(<AuthForm formType="login" />);

    expect(screen.getByRole("heading", { name: /login/i })).toBeInTheDocument();

    const registerNavButton = screen.getByRole("button", { name: /register/i });
    await userEvent.click(registerNavButton);

    expect(mockNavigate).toHaveBeenCalledWith("/register");
  });

  test("renders register form, allows submission, and redirects to login", async () => {

    const mockRegisterUser = jest.fn();
    (useAuthentication as jest.Mock).mockReturnValue({
      loginUser: jest.fn(),
      registerUser: mockRegisterUser,
    });

    render(<AuthForm formType="register" />);

    expect(
      screen.getByRole("heading", { name: /register/i })
    ).toBeInTheDocument();

    const nameInput = screen.getByLabelText(/name/i);
    const emailInput = screen.getByLabelText(/email/i);
    const passwordInput = screen.getByLabelText(/password/i);
    const roleSelect = screen.getByLabelText(/role/i);
    const ageInput = screen.getByLabelText(/age/i);
    const registerSubmitButton = screen.getByRole("button", {
      name: /register/i,
    });

    await userEvent.type(nameInput, "John Doe");
    await userEvent.type(emailInput, "john@example.com");
    await userEvent.type(passwordInput, "securePass123");
    await userEvent.selectOptions(roleSelect, "user");
    await userEvent.type(ageInput, "25");

    await userEvent.click(registerSubmitButton);

    expect(mockRegisterUser).toHaveBeenCalledWith({
        name:"John Doe",
        email:"john@example.com",
        password:"securePass123",
        role:"user",
        age:25
    })
    // expect(mockNavigate).toHaveBeenCalledWith("/login");
  });

  test("navigates back to login form when clicking login button", async () => {
    render(<AuthForm formType="register" />);

    expect(
      screen.getByRole("heading", { name: /register/i })
    ).toBeInTheDocument();

    const loginNavButton = screen.getByRole("button", { name: /login/i });
    await userEvent.click(loginNavButton);

    expect(mockNavigate).toHaveBeenCalledWith("/login");
  });
});
