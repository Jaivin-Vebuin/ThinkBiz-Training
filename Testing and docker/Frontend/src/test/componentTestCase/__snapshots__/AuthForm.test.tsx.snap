// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Form matches snapshot for login form 1`] = `
<DocumentFragment>
  <form
    class="sc-FEMpB jGlMdh"
  >
    <h1
      class="sc-gjcoXW jJIRPV"
    >
      formTitleLogin
    </h1>
    <div>
      <div
        class="sc-dVBluf hYubA-d"
      >
        <div>
          <label
            for="userEmail"
          >
            Email
          </label>
          <input
            id="userEmail"
            name="userEmail"
            placeholder="emailPlaceHolder"
            type="text"
          />
           
        </div>
      </div>
      <div
        class="sc-dVBluf hYubA-d"
      >
        <div>
          <label
            for="userPassword"
          >
            Password
          </label>
          <input
            id="userPassword"
            name="userPassword"
            placeholder="********"
            type="password"
          />
           
        </div>
      </div>
    </div>
    <div
      class="sc-fLDLck jlNlhs"
    >
      <button>
        loginFormButton
      </button>
    </div>
  </form>
  <div
    style="padding: 15px; border-radius: 8px; text-align: center; margin: 20px;"
  >
    forNewUser
    <button
      style="background: white; padding: 5px 10px; border-radius: 5px; font-weight: bold; cursor: pointer;"
    >
      registerFormButton
    </button>
  </div>
</DocumentFragment>
`;

exports[`Auth Form matches snapshot for register form 1`] = `
<DocumentFragment>
  <form
    class="sc-FEMpB jGlMdh"
  >
    <h1
      class="sc-gjcoXW jJIRPV"
    >
      formTitleRegister
    </h1>
    <div>
      <div
        class="sc-dVBluf hYubA-d"
      >
        <div>
          <label
            for="userName"
          >
            Name
          </label>
          <input
            id="userName"
            name="userName"
            placeholder="namePlaceHolder"
            type="text"
          />
           
        </div>
      </div>
      <div
        class="sc-dVBluf hYubA-d"
      >
        <div>
          <label
            for="userEmail"
          >
            Email
          </label>
          <input
            id="userEmail"
            name="userEmail"
            placeholder="emailPlaceHolder"
            type="text"
          />
           
        </div>
      </div>
      <div
        class="sc-dVBluf hYubA-d"
      >
        <div>
          <label
            for="userPassword"
          >
            Password
          </label>
          <input
            id="userPassword"
            name="userPassword"
            placeholder="********"
            type="password"
          />
           
        </div>
      </div>
      <div
        class="sc-dVBluf hYubA-d"
      >
        <div>
          <label
            for="userRole"
          >
            Role
          </label>
          <select
            id="userRole"
            name="userRole"
          >
             
            <option
              value="admin"
            >
              admin
            </option>
            ;
            <option
              value="user"
            >
              user
            </option>
            ;
          </select>
        </div>
      </div>
      <div
        class="sc-dVBluf hYubA-d"
      >
        <div>
          <label
            for="userAge"
          >
            Age
          </label>
          <input
            id="userAge"
            name="userAge"
            placeholder="agePlaceHolder"
            type="number"
          />
           
        </div>
      </div>
    </div>
    <div
      class="sc-fLDLck jlNlhs"
    >
      <button>
        registerFormButton
      </button>
    </div>
  </form>
  <div
    style="padding: 15px; border-radius: 8px; text-align: center; margin: 20px;"
  >
    forRegisteredUser
    <button
      style="background: white; padding: 5px 10px; border-radius: 5px; font-weight: bold; cursor: pointer;"
    >
      loginFormButton
    </button>
  </div>
</DocumentFragment>
`;
